File: JavaTest/src/main/java/com/tan/sort/Quick_Sort.java
package com.tan.sort;

import java.util.Arrays;

/**
 * Created by Tan on 2017/8/2.
 */
public class Quick_Sort {
    public static void main(String[] args) {
        int[] arr = {23,13,53,35,21,5,30};
        quick_sort(arr,0,arr.length-1);
        System.out.println(Arrays.toString(arr));
    }
    //   x=23     5,13,53,35,21,5,30
    //快速排序
    private static void quick_sort(int s[], int l, int r)
    {
        if (l < r)
        {
            //Swap(s[l], s[(l + r) / 2]); //将中间的这个数和第一个数交换 参见注1
            int i = l, j = r, x = s[l];
            while (i < j)
            {
                while(i < j && s[j] >= x) // 从右向左找第一个小于x的数
                    j--;
                if(i < j)
                    s[i++] = s[j];
                    System.out.println("i:"+i + "  j:"+j +"  "+Arrays.toString(s));

                while(i < j && s[i] < x) // 从左向右找第一个大于等于x的数
                    i++;
                if(i < j)
                    s[j--] = s[i];
                    System.out.println("i:"+i + "  j:"+j +"  "+Arrays.toString(s));
            }
            s[i] = x;
            quick_sort(s, l, i - 1); // 递归调用
            quick_sort(s, i + 1, r);
        }
    }
}

File: ScalaTest/src/QuickSort.scala
/**
  * Created by Tan on 2017/8/2.
  */
object QuickSort {
  def quickSort(list: List[Int]): List[Int] = {
    list match {
      case Nil => Nil
      case List() => List()
      case head :: tail =>
        val (left, right) = tail.partition(_ < head) //以将数组
        quickSort(left) ::: head :: quickSort(right)
    }
  }

  def main(args: Array[String]) {
    val list = List(3, 12, 43, 23, 7, 1, 2, 0)
    println("End: "+quickSort(list))
  }
//  def main(args: Array[String]): Unit = {
//    val arr:Array[Int]= Array(23,42,12,63,73,24,15)
//    sortModel(arr,0,arr.length-1)
//    println("ok")
//  }
//
//  def sortModel(arr: Array[Int], l: Int, r: Int) {
//    if(l<r){
//      var i=l
//      var j=r
//      var temp = arr(l)
//      while(i<j){
//        while(i<j && arr(j)>temp) j=j-1
//        if (i<j) arr(i+1)=arr(j)
//        while(i<j && arr(i)<temp) i+i+1
//        if (i<j) arr(j-1)=arr(i)
//      }
//      arr(i)=temp
//      sortModel(arr,l,j-1)
//      sortModel(arr,i+1,r)
//    }
//  }
}


